bitset.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * bitset.h -- Dynamic bitset.
+ *
+ * Copyright (c) 2001-2020, NLnet Labs. All rights reserved.
+ *
+ * See LICENSE for the license.
+ *
+ */
+
+#ifndef _BITSET_H_
+#define _BITSET_H_
+
+#include <assert.h>
+#include <limits.h>
+#include <string.h>
+
+typedef struct nsd_bitset nsd_bitset_type;
+
+struct nsd_bitset {
+	size_t size; /** Number of available bits in the set */
+	unsigned char bits[];
+};
+
+inline size_t
+nsd_bitset_size(size_t bits)
+{
+	if(bits == 0)
+		bits++;
+
+	return (bits / CHAR_BIT) + ((bits % CHAR_BIT) != 0) + sizeof(size_t);
+}
+
+inline void
+nsd_bitset_zero(struct nsd_bitset *bset)
+{
+	size_t sz;
+
+	assert(bset != NULL);
+
+	sz = nsd_bitset_size(bset->size) - sizeof(bset->size);
+	assert(sz > 0);
+	memset(bset->bits, 0, sz);
+}
+
+inline void
+nsd_bitset_init(struct nsd_bitset *bset, size_t bits)
+{
+	assert(bset != NULL);
+	if (bits == 0)
+		bits++;
+
+	bset->size = bits;
+	nsd_bitset_zero(bset);
+}
+
+inline int
+nsd_bitset_isset(struct nsd_bitset *bset, size_t bit)
+{
+	assert(bset != NULL);
+	if(bit >= bset->size)
+		return 0;
+
+	return (bset->bits[ (bit / CHAR_BIT) ] & (1 << (bit % CHAR_BIT))) != 0;
+}
+
+inline void
+nsd_bitset_set(struct nsd_bitset *bset, size_t bit)
+{
+	assert(bset != NULL);
+	assert(bset->size > bit);
+	bset->bits[ (bit / CHAR_BIT) ] |= (1 << (bit % CHAR_BIT));
+}
+
+inline void
+nsd_bitset_unset(struct nsd_bitset *bset, size_t bit)
+{
+	assert(bset != NULL);
+	assert(bset->size > bit);
+	bset->bits[ (bit / CHAR_BIT) ] &= ~(1 << (bit % CHAR_BIT));
+}
+
+inline void
+nsd_bitset_or(
+	struct nsd_bitset *destset,
+	struct nsd_bitset *srcset1,
+	struct nsd_bitset *srcset2)
+{
+	size_t size, bytes;
+	unsigned char bits, mask;
+
+	assert(destset != NULL);
+	assert(srcset1 != NULL);
+	assert(srcset2 != NULL);
+
+	size = destset->size;
+	bytes = (size / CHAR_BIT) + ((size & CHAR_BIT) != 0);
+
+	for(size_t i = 0, n; i < bytes; i++) {
+		bits = 0;
+
+		size = destset->size < srcset1->size ?
+		       destset->size : srcset1->size;
+		n = (size / CHAR_BIT);
+                if (n >= i) {
+			bits |= srcset1->bits[i];
+		} else {
+			n += ((size % CHAR_BIT) != 0);
+			mask = (1 << ((size % CHAR_BIT) + 1)) - 1;
```

Comment:
You need to make sure the 1<< is unsigned int, not unsigned char, that is the type of mask.  Otherwise you'd run out of bitspace for the shift.  It may be nice to make that explicit, with a cast.

Reply:
Correct